### PR TITLE
pre-spawn-environ

### LINKED ISSUE #6780: Set `MALLOC_TRIM_THRESHOLD_` before interpreter start
In https://github.com/dask/distributed/pull/6777 we reverted https://github.com/dask/distributed/pull/6681 in order to avoid https://github.com/dask/distributed/issues/6749 (which broke `dask-cuda`). This revert is meant to be temporary with the intent of restoring the functionality of https://github.com/dask/distributed/pull/6681, possibly addressing some of the implementation comments in https://github.com/dask/distributed/issues/6749. Opening up an issue so we don't loose track of getting https://github.com/dask/distributed/pull/6681's functionality back into `main`. 

cc @crusaderky @gjoseph92 @pentschev @jakirkham who all engaged on the previously linked to issue / PRs

### LINKED ISSUE #6833: Performance regression in cluster memory usage
Recently, some of us have started tracking some performance metrics for Dask clusters under a variety of different usage patterns. The idea is to be able to identify performance regressions before they are released (especially ones at scale which might not show up in a unit test context).

An example of these metrics is at [this](https://coiled.github.io/coiled-runtime/coiled-upstream-py3.9.html) static site. It's only been collecting results for a few days, but already we seem to have come across a significant regression in cluster memory usage.  Here is a test which measures array rechunking:
```python
def test_rechunk_in_memory(small_client):
    x = da.random.random((50000, 50000))
    x.rechunk((50000, 20)).rechunk((20, 50000)).sum().compute()
```
and a screenshot of average cluster memory usage for that operation over the last week+ :
image
(I encourage folks to click through, this same behavior appears on a lot of tests around July 26)

The above is based on a coiled cluster, but I've reproduced it using a `LocalCluster` with the following procedure:

1. Create a software environment with nightly dask versions from the `dask` conda channel:
    ```bash
    conda create -n memory-regression python=3.9 dask distributed numpy
    conda activate memory-regression
    # Install nightly from July 22nd
    conda install https://conda.anaconda.org/dask/label/dev/noarch/dask-2022.7.1a220722-py_ga55bfd36_21.tar.bz2 https://conda.anaconda.org/dask/label/dev/noarch/distributed-2022.7.1a220722-py_ga55bfd36_21.tar.bz2
   # Or install nightly from July 25th
   conda install https://conda.anaconda.org/dask/label/dev/noarch/dask-2022.7.2a220725-py_g55cc1a50_1.tar.bz2 https://conda.anaconda.org/dask/label/dev/noarch/distributed-2022.7.2a220725-py_g55cc1a50_1.tar.bz2
    ```
2. Run the following script
```python
import ctypes
import uuid

import dask.array as da
import distributed


sampler = distributed.diagnostics.MemorySampler()


def trim_memory() -> int:
    libc = ctypes.CDLL("libc.so.6")
    return libc.malloc_trim(0)


if __name__ == "__main__":
    print(distributed.__version__)
    client = distributed.Client()
    mems = []
    for i in range(20):
        label = str(uuid.uuid4())

        with sampler.sample(label=label, client=client, measure="process"):
            x = da.random.random((20000, 20000))
            x.rechunk((20000, 20)).rechunk((20, 20000)).sum().compute()

        df = sampler.to_pandas()
        mems.append(df[label].mean())
        client.run(trim_memory)
        client.restart()

    print(mems)
```

This produces results like the following:

image

~~Timing-wise, this suggests to me that #6728 might have had some unintended side-effects in cluster memory usage, but I have not verified, nor do I know how it could be so drastic.~~
Edit, see below

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes #6780
- Closes #6833

### CODE REVIEW

File: distributed/nanny.py [hunk 1]
```diff
@@ -82,12 +83,27 @@ class Nanny(ServerNode):
...
             2. Existing environment variables
             3. Dask configuration
 
+        Note
+        ----
+        Some environment variables, like ``OMP_NUM_THREADS``, must be set before
+        importing numpy to have effect. Others, like ``MALLOC_TRIM_THRESHOLD_`` (see
+        :ref:`memtrim`), must be set before starting the Linux process. Such variables
+        would be ineffective if set here or in ``distributed.nanny.environ``; they
+        must be set in ``distributed.nanny.pre-spown-environ`` so that they are set
```
Comment (fjetter) [inline distributed/nanny.py:98]:
```suggestion
        must be set in ``distributed.nanny.pre-spawn-environ`` so that they are set
```

**Review by fjetter — APPROVED** [review]
Changes LGTM but I think there is a genuine test failure in test_config about the schema validation

**Comment by crusaderky** [comment]
fixed test

File: distributed/distributed-schema.yaml [hunk 2]
```diff
@@ -568,7 +568,14 @@ properties:
...
             description: |
-              Environment variables to set on all worker processes started by nannies
+              Environment variables to set on all worker processes started by nannies.
+              These variables are set after spawning the worker process.
+
+          pre-spawn-environ:
+            type: object
+            description: |
+              Environment variables to set on all worker processes started by nannies.
+              These variables are set before spawning the worker process.
```
Comment (gjoseph92) [inline distributed/distributed-schema.yaml:578]:
```suggestion
              These variables are set within the Nanny process,
              before spawning the worker process.
              Should be used for variables that must be set before
              process startup, interpreter startup, or imports.
```

File: distributed/tests/test_nanny.py [hunk 3]
```diff
@@ -330,6 +330,25 @@ async def test_environment_variable_config(c, s, monkeypatch):
...
+        "distributed.nanny.environ": {"POST-SPAWN": 2},
+    },
+)
+async def test_environment_variable_pre_post_spawn(c, s, n):
+    assert n.env == {"PRE-SPAWN": "1", "POST-SPAWN": "2"}
+    results = await c.run(lambda: os.environ)
+    assert results[n.worker_address]["PRE-SPAWN"] == "1"
+    assert results[n.worker_address]["POST-SPAWN"] == "2"
+
+    del os.environ["PRE-SPAWN"]
```
Comment (gjoseph92) [inline distributed/tests/test_nanny.py:348]: in a `finally`? not a big deal if it leaks to other tests though

Comment (crusaderky) [reply distributed/tests/test_nanny.py:348]: a bit overkill IMHO

**Comment by jrbourbeau** [comment]
Thanks @crusaderky

cc @pentschev if you want to try out the current `main` branch with `dask-cuda`

**Comment by pentschev** [comment]
Thanks @crusaderky for the work here and @jrbourbeau for the ping. On initial testing, this PR seems to be working well for us. We should have some heavier testing done overnight and if anything pops up, I'll ping this thread again, but hopefully won't be necessary!

**Pull request merged**